walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,80 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    int minutesInOneDay = 24 * 60;
+    if (request.getDuration() > minutesInOneDay)
+        return new ArrayList<TimeRange>();
+
+    Collection<TimeRange> withOptional = new ArrayList<>();
+    Collection<TimeRange> mandatoryOnly = new ArrayList<>();
+    withOptional.add(TimeRange.WHOLE_DAY);
+    mandatoryOnly.add(TimeRange.WHOLE_DAY);
+
+    Collection<String> allAttendees = new ArrayList<>();
+    allAttendees.addAll(request.getAttendees());
+    allAttendees.addAll(request.getOptionalAttendees());
+    for (Event event : events) {
+        if (!areAttendeesAvailable(allAttendees, event))
+        {
+            withOptional = availableTimes(withOptional, event, request.getDuration());
+        }
+        if (!areAttendeesAvailable(request.getAttendees(), event))
+        {
+            mandatoryOnly = availableTimes(mandatoryOnly, event, request.getDuration());
+        }
+    }
+
+    if (request.getAttendees().isEmpty() || !withOptional.isEmpty())
```

Comment:
Would this work without the `request.getAttendees().isEmpty()`? If there are only optional attendees and nothing fits for the optional attendees, should we be returning nothing is available or everything is available?

Reply:
one of the weird test cases they had was if a meeting has no attendees (mandatory or optional), then you should return that the whole day is available, but if you have no mandatory attendees but some optional attendees, you should return the times that work for the optional attendees even if no times work. so that's why i added that logic there